cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,869 @@
+
+    /**
+     * @brief Compute the offset to the end of the window.
+     *
+     * @param i The current row index.
+     * @return Offset to the current row's window endpoint.
+     */
+    [[nodiscard]] __device__ size_type operator()(size_type i) const
+    {
+      using Comp      = comparator_t<OrderbyT, WindowTag, Order>;
+      auto const info = groups.row_info(i);
+      auto const [null_count, group_start, group_end, null_start, null_end, start, end] =
+        groups.row_info(i);
+      if constexpr (Direction == direction::PRECEDING) {
+        if constexpr (WindowTag == window_tag::UNBOUNDED) { return i - group_start + 1; }
+        // TODO: If the window is BOUNDED_OPEN, what does it mean for a row to fall in the null
+        // group? Not that important because only spark allows nulls in the orderby column, and it
+        // doesn't have BOUNDED_OPEN windows.
+        if (Grouping::has_nulls && i >= null_start && i < null_end) { return i - null_start + 1; }
+        if constexpr (WindowTag == window_tag::CURRENT_ROW) {
+          return 1 + thrust::distance(
+                       thrust::lower_bound(thrust::seq, begin + start, begin + i, begin[i], Comp{}),
+                       begin + i);
+        } else if constexpr (WindowTag != window_tag::UNBOUNDED) {
+          // The preceding endpoint is computed via row_value - delta.
+          // When delta is positive, this can only overflow towards -infinity.
+          // If we did overflow towards -infinity, then the value
+          // we're searching for is some min. But -infinity < min, so
+          // we must always use a `BOUNDED_CLOSED` window so that
+          // orderby = [min, ...] with positive delta picks up that
+          // row in the window.
+          // Conversely, when delta is negative, we can only overflow
+          // towards +infinity. If we did overflow towards +infinity
+          // then the value we're searching for is some max. But
+          // +infinity > max, so we must use a `BOUNDED_OPEN` window
+          // so that orderby = [..., max] with negative delta does not
+          // pick up that row in the window.
+          // When the orderby column is sorted in descending order the
+          // above applies mutatis mutandis with a sign flip.
+          OrderbyT value;
+          bool did_overflow{false};
+          bool delta_positive{*row_delta > DeltaT{0}};
+          if constexpr (Order == order::ASCENDING) {
+            auto const result = saturating_sub(begin[i], *row_delta);
+            value             = cuda::std::get<0>(result);
+            did_overflow      = cuda::std::get<1>(result);
+          } else {
+            auto const result = saturating_add(begin[i], *row_delta);
+            value             = cuda::std::get<0>(result);
+            did_overflow      = cuda::std::get<1>(result);
+          }
+          if (did_overflow) {
```

Comment:
I suppose one could avoid the declaration without initialisation by writing this as a lambda.